packages/contracts/src/types/index.ts

```diff
@@ -0,0 +1,56 @@
+import { BytesLike, Weight } from '@dedot/codecs';
+import { AnyFunc, AsyncMethod, GenericSubstrateApi } from '@dedot/types';
+import { ContractMessage } from './shared.js';
+import { ContractMetadataV4 } from './v4.js';
+import { ContractMetadataV5 } from './v5.js';
+
+export * from './shared.js';
+
+export type ContractMetadata = ContractMetadataV4 | ContractMetadataV5;
+
+export type ContractResult<ChainApi extends GenericSubstrateApi> = Awaited<
+  ReturnType<ChainApi['call']['contractsApi']['call']>
+>;
+
+// Now we are using this one for api.tx.contract.instantiate, api.tx.contract.instantiateWithCode and api.tx.contract.call
+// TODO: Write types for api.tx.contract.instantiate and api.tx.contract.instantiateWithCode
+export type ChainSubmittableExtrinsic<ChainApi extends GenericSubstrateApi> = ReturnType<
+  ChainApi['tx']['contracts']['call']
+>;
+
+export type ConstructorOptions = ContractOptions & {
+  salt: BytesLike | null;
+};
+
+export type ContractOptions = {
+  value: bigint;
+  gasLimit: Weight | undefined;
+  storageDepositLimit: bigint | undefined;
+};
+
+export interface GenericContractResult<DecodedData, ContractResult> {
+  data?: DecodedData;
```

Comment:
Can this `data` sometime return `undefined`?
```suggestion
  data: DecodedData;
```